cli/internal/core/engine_persistent_deps_test.go

```diff
@@ -0,0 +1,231 @@
+package core
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+
+	testifyAssert "github.com/stretchr/testify/assert"
+	"github.com/vercel/turbo/cli/internal/fs"
+	"github.com/vercel/turbo/cli/internal/graph"
+	"github.com/vercel/turbo/cli/internal/util"
+	"gotest.tools/v3/assert"
+
+	"github.com/pyr-sh/dag"
+)
+
+var _workspaceGraphDefinition = map[string][]string{
+	"workspace-a": {"workspace-c"}, // a depends on c
+	"workspace-b": {"workspace-c"}, // b depends on c
+	"workspace-c": {},
+}
+
+func TestPrepare_PersistentDependencies_Topological(t *testing.T) {
+	completeGraph, workspaces := _buildCompleteGraph(_workspaceGraphDefinition)
+
+	engine := NewEngine(&completeGraph.TopologicalGraph)
+
+	// "dev": dependsOn: ["^dev"] (where dev is persistent)
+	engine.AddTask(&Task{
+		Name:       "dev",
+		TopoDeps:   util.SetFromStrings([]string{"dev"}),
+		Deps:       make(util.Set), // empty, no non-caret task deps.
+		Persistent: true,
+	})
+
+	opts := &EngineBuildingOptions{
+		Packages:  workspaces,
+		TaskNames: []string{"dev"},
+	}
+
+	err := engine.Prepare(opts)
+	assert.NilError(t, err, "Failed to prepare engine")
+
+	// do the validation
+	actualErr := engine.ValidatePersistentDependencies(completeGraph)
+
+	// Use a regex here, because depending on the order the graph is walked,
+	// either workpsce-a or workspace-b could throw the error first.
+	expected := regexp.MustCompile("\"workspace-c#dev\" is a persistent task, \"workspace-[a|b]#dev\" cannot depend on it")
+	testifyAssert.Regexp(t, expected, actualErr)
+}
+
+func TestPrepare_PersistentDependencies_SameWorkspace(t *testing.T) {
```

Comment:
You know what would be amazing here? A microformat-reading graph creator so that we could quickly specify really complicated graphs in an understandable way.

But that's a bad idea. As a compromise can you draw out the graph for these in ascii art? (I typically use `mkdir` and `tree` to do this.)

Reply:
I can add some diagrams. The package dependencies graph is drawn out in a more understandable format at the top of the file